File: answers/10.Simulation/distribution_simulation.jl
# Distribution Simulation in Julia
# Simulating random variables from various probability distributions

using Distributions
using Random
using Plots
using Statistics

println("=== Distribution Simulation ===\n")

# Set seed for reproducibility
Random.seed!(42)

# Key Concept: Four types of distribution functions
println("--- Four Distribution Functions ---")
println("1. pdf/pmf: Probability density/mass function")
println("2. cdf: Cumulative distribution function")
println("3. quantile: Inverse CDF (percentile)")
println("4. rand: Generate random samples\n")

# Example with Normal distribution
println("Example: Normal Distribution N(μ=5, σ=2)")
μ, σ = 5, 2
x = 6

println("\n1. PDF at x=$x: ", round(pdf(Normal(μ, σ), x), digits=6))
println("2. CDF P(X ≤ $x): ", round(cdf(Normal(μ, σ), x), digits=6))
println("3. 95th percentile: ", round(quantile(Normal(μ, σ), 0.95), digits=6))
println("4. Random sample (n=5): ", rand(Normal(μ, σ), 5))

# Common distributions and their simulation
println("\n--- Simulating Common Distributions ---\n")

n_samples = 1000

# 1. Normal
normal_samples = rand(Normal(μ, σ), n_samples)
println("1. Normal(5, 2): mean=$(round(mean(normal_samples), digits=3)), std=$(round(std(normal_samples), digits=3))")

# 2. Exponential
exp_samples = rand(Exponential(2.0), n_samples)  # scale=2 means rate=0.5
println("2. Exponential(λ=0.5): mean=$(round(mean(exp_samples), digits=3))")

# 3. Binomial
binom_samples = rand(Binomial(10, 0.3), n_samples)
println("3. Binomial(n=10, p=0.3): mean=$(round(mean(binom_samples), digits=3))")

# 4. Poisson
poisson_samples = rand(Poisson(4.5), n_samples)
println("4. Poisson(λ=4.5): mean=$(round(mean(poisson_samples), digits=3))")

# 5. Uniform
uniform_samples = rand(Uniform(0, 1), n_samples)
println("5. Uniform(0, 1): mean=$(round(mean(uniform_samples), digits=3))")

# Visualization
p1 = histogram(normal_samples, bins=30, normalize=:pdf, label="Data",
               xlabel="Value", ylabel="Density", title="Normal(5, 2)", alpha=0.7)
x_range = range(minimum(normal_samples), maximum(normal_samples), length=100)
plot!(p1, x_range, pdf.(Normal(μ, σ), x_range), linewidth=2, label="PDF", color=:red)

p2 = histogram(exp_samples, bins=30, normalize=:pdf, label="Data",
               xlabel="Value", title="Exponential(λ=0.5)", alpha=0.7)
x_range = range(0, maximum(exp_samples), length=100)
plot!(p2, x_range, pdf.(Exponential(2.0), x_range), linewidth=2, label="PDF", color=:red)

p3 = histogram(binom_samples, bins=-0.5:1:10.5, normalize=:pdf, label="Data",
               xlabel="Value", title="Binomial(n=10, p=0.3)", alpha=0.7)
x_vals = 0:10
plot!(p3, x_vals, pdf.(Binomial(10, 0.3), x_vals), marker=:circle,
      linewidth=2, label="PMF", color=:red)

p4 = histogram(poisson_samples, bins=-0.5:1:14.5, normalize=:pdf, label="Data",
               xlabel="Value", title="Poisson(λ=4.5)", alpha=0.7)
x_vals = 0:14
plot!(p4, x_vals, pdf.(Poisson(4.5), x_vals), marker=:circle,
      linewidth=2, label="PMF", color=:red)

p5 = histogram(uniform_samples, bins=30, normalize=:pdf, label="Data",
               xlabel="Value", title="Uniform(0, 1)", alpha=0.7, ylims=(0, 1.5))
hline!(p5, [1], linewidth=2, label="PDF", color=:red)

p6 = scatter(quantile.(Normal(), (1:n_samples) ./ (n_samples+1)),
             sort(normal_samples), label="", xlabel="Theoretical Quantiles",
             ylabel="Sample Quantiles", title="Q-Q Plot (Normal)", alpha=0.5)
plot!(p6, [-3, 3], [-3, 3], linewidth=2, linestyle=:dash, color=:red, label="")

plot(p1, p2, p3, p4, p5, p6, layout=(2, 3), size=(1500, 1000))
savefig("/home/titan/pdfs/notes/statisticalComputingAndReporting/groupWork/answers/10.Simulation/distribution_simulation_jl.png")
println("\nVisualization saved as distribution_simulation_jl.png")

File: answers/10.Simulation/distribution_simulation.py
"""
Distribution Simulation
Simulating random variables from various probability distributions
"""
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats

print("=== Distribution Simulation ===\n")

# Set seed for reproducibility
np.random.seed(42)

# Key Concept: Four types of distribution functions
print("--- Four Distribution Functions ---")
print("1. pdf/pmf (d): Probability density/mass function")
print("2. cdf (p): Cumulative distribution function")
print("3. quantile (q): Inverse CDF (percentile)")
print("4. random (r): Generate random samples\n")

# Example with Normal distribution
print("Example: Normal Distribution N(μ=5, σ=2)")
mu, sigma = 5, 2

x = 6
print(f"\n1. PDF at x={x}: {stats.norm.pdf(x, mu, sigma):.6f}")
print(f"2. CDF P(X ≤ {x}): {stats.norm.cdf(x, mu, sigma):.6f}")
print(f"3. 95th percentile: {stats.norm.ppf(0.95, mu, sigma):.6f}")
print(f"4. Random sample (n=5): {stats.norm.rvs(mu, sigma, size=5)}")

# Common distributions and their simulation
print("\n--- Simulating Common Distributions ---\n")

n_samples = 1000

# 1. Normal
normal_samples = np.random.normal(mu, sigma, n_samples)
print(f"1. Normal(5, 2): mean={np.mean(normal_samples):.3f}, std={np.std(normal_samples):.3f}")

# 2. Exponential
exp_samples = np.random.exponential(scale=2.0, size=n_samples)
print(f"2. Exponential(λ=0.5): mean={np.mean(exp_samples):.3f}")

# 3. Binomial
binom_samples = np.random.binomial(n=10, p=0.3, size=n_samples)
print(f"3. Binomial(n=10, p=0.3): mean={np.mean(binom_samples):.3f}")

# 4. Poisson
poisson_samples = np.random.poisson(lam=4.5, size=n_samples)
print(f"4. Poisson(λ=4.5): mean={np.mean(poisson_samples):.3f}")

# 5. Uniform
uniform_samples = np.random.uniform(0, 1, n_samples)
print(f"5. Uniform(0, 1): mean={np.mean(uniform_samples):.3f}")

# Visualization
fig, axes = plt.subplots(2, 3, figsize=(15, 10))

# Plot 1: Normal
axes[0, 0].hist(normal_samples, bins=30, density=True, alpha=0.7, edgecolor='black')
x_range = np.linspace(normal_samples.min(), normal_samples.max(), 100)
axes[0, 0].plot(x_range, stats.norm.pdf(x_range, mu, sigma), 'r-', linewidth=2)
axes[0, 0].set_title('Normal(5, 2)')
axes[0, 0].set_xlabel('Value')
axes[0, 0].set_ylabel('Density')

# Plot 2: Exponential
axes[0, 1].hist(exp_samples, bins=30, density=True, alpha=0.7, edgecolor='black')
x_range = np.linspace(0, exp_samples.max(), 100)
axes[0, 1].plot(x_range, stats.expon.pdf(x_range, scale=2.0), 'r-', linewidth=2)
axes[0, 1].set_title('Exponential(λ=0.5)')
axes[0, 1].set_xlabel('Value')

# Plot 3: Binomial
axes[0, 2].hist(binom_samples, bins=range(12), density=True, alpha=0.7, edgecolor='black', align='left')
x_vals = range(11)
axes[0, 2].plot(x_vals, stats.binom.pmf(x_vals, 10, 0.3), 'ro-', linewidth=2)
axes[0, 2].set_title('Binomial(n=10, p=0.3)')
axes[0, 2].set_xlabel('Value')

# Plot 4: Poisson
axes[1, 0].hist(poisson_samples, bins=range(15), density=True, alpha=0.7, edgecolor='black', align='left')
x_vals = range(15)
axes[1, 0].plot(x_vals, stats.poisson.pmf(x_vals, 4.5), 'ro-', linewidth=2)
axes[1, 0].set_title('Poisson(λ=4.5)')
axes[1, 0].set_xlabel('Value')

# Plot 5: Uniform
axes[1, 1].hist(uniform_samples, bins=30, density=True, alpha=0.7, edgecolor='black')
axes[1, 1].axhline(y=1, color='r', linewidth=2)
axes[1, 1].set_title('Uniform(0, 1)')
axes[1, 1].set_xlabel('Value')
axes[1, 1].set_ylim([0, 1.5])

# Plot 6: Q-Q plot for normality check
axes[1, 2].scatter(*stats.probplot(normal_samples, dist="norm")[0], alpha=0.5)
axes[1, 2].plot([-3, 3], [-3, 3], 'r--', linewidth=2)
axes[1, 2].set_title('Q-Q Plot (Normal samples)')
axes[1, 2].set_xlabel('Theoretical Quantiles')
axes[1, 2].set_ylabel('Sample Quantiles')

plt.tight_layout()
plt.savefig('/home/titan/pdfs/notes/statisticalComputingAndReporting/groupWork/answers/10.Simulation/distribution_simulation.png', dpi=150)
print("\nVisualization saved as distribution_simulation.png")
plt.close()
